### PR TITLE
WCCP: Fix GCC-10 -Wstringop-truncation failures

### PR DESCRIPTION
**Pull request by sergiodj**
I can only trigger these failures when I compile on s390x.

    In function 'char* strncpy(char*, const char*, size_t)', output
        may be truncated copying 8 bytes from a string of length 8

Signed-off-by: Sergio Durigan Junior <sergiodj@debian.org>

### CODE REVIEW

**Comment by kinkie** [comment]
ok to test

File: src/wccp2.cc [hunk 1]
```diff
@@ -553,17 +552,16 @@ static char
...
     struct wccp2_security_md5_t *ws;
 
     debugs(80, 5, "wccp2_update_md5_security: called");
 
     /* The password field, for the MD5 hash, needs to be 8 bytes and NUL padded. */
-    memset(pwd, 0, sizeof(pwd));
-    strncpy(pwd, password, sizeof(pwd));
-    pwd[sizeof(pwd) - 1] = '\0';
+    memset(pwd, 0, WCCP2_PASSWORD_LEN);
+    xstrncpy(pwd, password, WCCP2_PASSWORD_LEN);
```
Comment (yadij) [inline src/wccp2.cc:564]: The sizeof(pwd) were correct. Please revert the change to size parameters for both these calls.

```suggestion
    memset(pwd, 0, sizeof(pwd));  // prevent WCCP leaking data from stack area
    xstrncpy(pwd, password, sizeof(pwd));
```

File: src/wccp2.cc [hunk 2]
```diff
@@ -629,9 +627,8 @@ wccp2_check_security(struct wccp2_service_list_t *srv, char *security, char *pac
     /* If execution makes it here then we have an MD5 security */
 
     /* The password field, for the MD5 hash, needs to be 8 bytes and NUL padded. */
-    memset(pwd, 0, sizeof(pwd));
-    strncpy(pwd, srv->wccp_password, sizeof(pwd));
-    pwd[sizeof(pwd) - 1] = '\0';
+    memset(pwd, 0, WCCP2_PASSWORD_LEN);
+    xstrncpy(pwd, srv->wccp_password, WCCP2_PASSWORD_LEN);
```
Comment (yadij) [inline src/wccp2.cc:631]: Again here, please leave the sizeof(pwd) as the size parameter.

File: src/wccp2.cc [hunk 3]
```diff
@@ -519,8 +519,7 @@ wccp2_add_service_list(int service, int service_id, int service_priority,
     wccp2_update_service(wccp2_service_list_ptr, service, service_id,
                          service_priority, service_proto, service_flags, ports);
     wccp2_service_list_ptr->wccp2_security_type = security_type;
-    memset(wccp2_service_list_ptr->wccp_password, 0, WCCP2_PASSWORD_LEN + 1);
-    strncpy(wccp2_service_list_ptr->wccp_password, password, WCCP2_PASSWORD_LEN);
+    xstrncpy(wccp2_service_list_ptr->wccp_password, password, WCCP2_PASSWORD_LEN);
```
Comment (yadij) [inline src/wccp2.cc:522]: This will truncate 1 character off the end of passwords that fill the buffer because the n parameter to xstrncpy() includes the c-string terminator character.

For these conversions you need to use the length parameter from the memset() as buffer size instead of the pre-shortened one from the old strncpy().

File: src/wccp2.cc [hunk 4]
```diff
@@ -2137,7 +2134,7 @@ parse_wccp2_service(void *)
     if ((t = ConfigParser::NextToken()) != NULL) {
         if (strncmp(t, "password=", 9) == 0) {
             security_type = WCCP2_MD5_SECURITY;
-            strncpy(wccp_password, t + 9, WCCP2_PASSWORD_LEN);
+            xstrncpy(wccp_password, t + 9, WCCP2_PASSWORD_LEN);
```
Comment (yadij) [inline src/wccp2.cc:2137]: Please use sizeof() for the size parameter here to prevent string truncation.
```suggestion
            xstrncpy(wccp_password, t + 9, sizeof(wccp_password));
```

File: src/wccp2.cc (hunk 1 above)
Comment (sergiodj) [reply src/wccp2.cc]: Done.

File: src/wccp2.cc (hunk 2 above)
Comment (sergiodj) [reply src/wccp2.cc]: Done.

File: src/wccp2.cc (hunk 3 above)
Comment (sergiodj) [reply src/wccp2.cc]: You're correct, thanks.  Fixed.

File: src/wccp2.cc (hunk 4 above)
Comment (sergiodj) [reply src/wccp2.cc]: Done.

**Comment by sergiodj** [comment]
Sorry about the mistakes; I was experimenting with `strncat` before finding `xstrncpy`, and apparently got confused by the buffer size parameter.

I force-pushed the branch with the fixes now.  Please let me know what you think.

**Comment by rousskov** [comment]
Thank you for fixing this bug! I have adjusted PR description to be more suitable for the anticipated automated [commit message](https://github.com/measurement-factory/anubis/blob/master/README.md#commit-message). Please polish further as needed.

**Comment by yadij** [comment]
If you want to go further and polish this up better all the +1 can be moved to the initial #define statement. I do not insist on this change.
```
#define WCCP2_PASSWORD_LEN  8+1  /* +1 for c-string nul terminator */
```

**Comment by sergiodj** [comment]
On Saturday, August 08 2020, Amos Jeffries wrote:

> If you want to go further and polish this up better all the +1 can be moved to the initial #define statement. I do not insist on this change.
> ```
> #define WCCP2_PASSWORD_LEN  8+1  /* +1 for c-string nul terminator */
> ```

Sure, no problem.  I force-pushed with the proposed change.  Let me know
what you think.

Thanks,

-- 
Sergio
GPG key ID: 237A 54B1 0287 28BF 00EF  31F4 D0EB 7628 65FC 5E36
Please send encrypted e-mail if possible
https://sergiodj.net/

**Comment by yadij** [comment]
Cleared for merge under 10-day no objections policy.
